### ISSUE TITLE
Computing gradients for custom optimization routine (and actor-critic RL methods)

### ISSUE THREAD

**Issue by aravindr93** [fchollet/keras#3062]
I would like to compute the gradients wrt weights of the neural network for studying a custom optimization procedure. This issue will also be useful for people trying to compute non-standard gradients (e.g. actor-critic algorithms in deep RL).

The gist of the scenario is described below in tensor-flow, but is easily translatable to Theano as well:
```
model = Sequential()
model.add(....) # add a couple of layers here
x = tf.placeholder(tf.float32, shape=(None, input_dim))
y = model(x)
loss = K.sum(K.square(y-target))  # just think of any standard loss fn
```
I need the gradient of loss wrt each parameter in the neural network and don't want to use the standard `model.fit()` function. Roughly, what i want to accomplish is:
```
param_grad_sym = tf.gradients(loss, model_params)  # symbolic
param_grad = sess.run(param_grad_sym, \
                       feed_dict={x: input_vals, model_params: model.get_weights()})
new_weights = custom_optimization_routine(param_grad, model.get_weights(), other_args)
model.set_weight(new_weights)
```

The problem is, I am unable to get the symbolic model_params required for tf.gradients or K.gradient. model.get_weights() returns the numerical values, but I need the symbolic variables. Thanks!

**Comment by ncullen93** [comment]
see my recent pull request -- https://github.com/fchollet/keras/pull/3064/files

Regardless of whether it's accepted, it certainly does what you need I think.

**Comment by aravindr93** [comment]
@ncullen93 Thanks! Issue resolved. The solution is along the following lines:
```
import keras
def get_trainable_params(model):
    params = []
    for layer in model.layers:
        params += keras.engine.training.collect_trainable_weights(layer)
    return params
```
Following this, the gradients can be computed using:
```
network_params = get_trainable_params(model)
param_grad = tf.gradients(cost, network_params)
```

For Actor-Critic RL models, one can compute the policy gradient (update equations for actor network) using this as follows:
```
# first define the model that takes state as input and outputs actions
action = model(state)
param_grad = tf.gradients(action, network_params, grad_ys=dQ_da)
# where dQ_da is grad of critic wrt action
```

**Comment by aravindr93** [comment]
@ncullen93 The code you shared isn't quite correct. The issue is summarized in #3069 , though I have mentioned tensorflow explicitly, the issue is also true for `K.get_value()` you have used in `pack_theta` function.

The issue in OP was resolved, so I am leaving this closed.

**Comment by jnhelen** [comment]
@aravindr93  Hi I am encountering the same problem and after looking your code below, it is still confusing me, I just can not pass the weight(which is defined in Keras) to the compute_gradients function in TF as the graph in the TF seems got incompleted. Could you help me with the problem? That's really important to me! Thank you!!!
